### PR TITLE
[SPARK-110] Support multiple spark dispatchers

### PR DESCRIPTION
**Pull request by tnachen**
Allow the Dcos spark cli to switch between different Spark dispatchers, which I introduced the concept of a 'app'. For now this is just a way to switch the targetted Spark in marathon based on the app id.
This is tested with the custom spark-build patch and with @s-urbaniak patch in Spark.

### CODE REVIEW

**Comment by mgummelt** [comment]
@tnachen please mark your PRs w/ JIRA numbers

**Comment by tnachen** [comment]
Merging this for now as the new admin router changes require this.

File: dcos_spark/discovery.py [hunk 1]
```diff
@@ -18,4 +19,4 @@ def get_spark_dispatcher():
     base_url = util.get_config().get('core.dcos_url')
     # Remove http:// prefix.
 
-    return base_url[7:] + '/service/sparkcli/'
+    return base_url[7:] + '/service/' + service.app_id() + '/'
```
Comment (mgummelt) [inline dcos_spark/discovery.py]: The CLI is at "/v1", not "/", right? https://github.com/mesosphere/spark-build/pull/9/files#diff-f5849f4ca7387dc9d1b4a1099222291cR10

File: dcos_spark/data/config-schema/spark.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  "properties": {
+    "app_id": {
+      "type": "string",
+      "title": "Spark App ID",
+      "description": "App ID that the Spark CLI is targetting to.",
+      "default": "spark"
+    }
+  },
+  "additionalProperties": false
+}
```
Comment (mgummelt) [inline dcos_spark/data/config-schema/spark.json]: newline at end of file

File: dcos_spark/data/config-schema/spark.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "$schema": "http://json-schema.org/schema#",
+  "type": "object",
+  "properties": {
+    "app_id": {
+      "type": "string",
+      "title": "Spark App ID",
+      "description": "App ID that the Spark CLI is targetting to.",
```
Comment (mgummelt) [inline dcos_spark/data/config-schema/spark.json]: "Marathon App ID of the Spark Dispatcher."

File: dcos_spark/cli.py [hunk 4]
```diff
@@ -80,13 +81,16 @@ def print_webui(args):
...
+    schema = pkg_resources.resource_string(
+        'dcos_spark',
+        'data/config-schema/spark.json').decode('utf-8')
+    print(schema)
 
 
 def main():
     args = docopt.docopt(
         __doc__,
-        version='dcos-spark version {}'.format(config.version), help=False)
```
Comment (mgummelt) [inline dcos_spark/cli.py]: With this rename, you need to update the reference here: https://github.com/mesosphere/dcos-spark/blob/master/bin/publish_to_s3.sh

File: dcos_spark/service.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from dcos import config, util
+
```
Comment (mgummelt) [inline dcos_spark/service.py]: This file should probably be called `app.py`

File: dcos_spark/spark_submit.py [hunk 6]
```diff
@@ -37,7 +37,8 @@ def spark_app():
     client = marathon.create_client()
     apps = client.get_apps()
     for marathon_app in apps:
-        if marathon_app.get('labels', {}).get('DCOS_PACKAGE_NAME') == 'spark':
+        if marathon_app.get('labels', {}).get('DCOS_PACKAGE_FRAMEWORK_NAME') \
```
Comment (mgummelt) [inline dcos_spark/spark_submit.py]: Why not compare to the app_id?  What is DCOS_PACKAGE_FRAMEWORK_NAME?

File: dcos_spark/data/config-schema/spark.json (hunk 3 above)
Comment (mgummelt) [inline dcos_spark/data/config-schema/spark.json]: Oh, it's not necessarily the Marathon App ID.  It's the framework-name: https://github.com/mesosphere/spark-build/pull/10/files

Then this should be "The dispatcher's framework name.  This is set via the `spark.framework-name` configuration option in the spark DCOS package."

We should also rename "app_id" to "framework_name"

File: dcos_spark/service.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+def app_id():
+    try:
+        return util.get_config()["spark.app_id"]
+    except KeyError:
+        return "spark"
+
+
+def set_app_id(app_id):
```
Comment (mgummelt) [inline dcos_spark/service.py]: This is never used.

File: dcos_spark/discovery.py (hunk 1 above)
Comment (tnachen) [inline dcos_spark/discovery.py]: /v1 is what spark submit will end up calling, but we need to pass just the correct URL prefix to spark submit.

File: dcos_spark/service.py (hunk 5 above)
Comment (tnachen) [inline dcos_spark/service.py]: I originally called it that, but it then clashed with dcos.app. I'm not quite familiar with python to know how to resolve this.

File: dcos_spark/spark_submit.py (hunk 6 above)
Comment (tnachen) [inline dcos_spark/spark_submit.py]: it is comparing to app_id which is service.app_id().
DCOS_PACKAGE_NAME is always spark regardless what framework_name you choose to install, but we need to support multiple spark dispatchers.

File: dcos_spark/discovery.py (hunk 1 above)
Comment (mgummelt) [inline dcos_spark/discovery.py]: ah ok.  I think the "/v1" in the proxy should be renamed to "dispatcher_cli/v1" or something similar.

File: dcos_spark/service.py (hunk 5 above)
Comment (mgummelt) [inline dcos_spark/service.py]: `from dcos import app as dcos_app`

File: dcos_spark/spark_submit.py (hunk 6 above)
Comment (mgummelt) [inline dcos_spark/spark_submit.py]: Nevermind.  I misread this.

**Pull request merged**